server/api.go

```diff
@@ -0,0 +1,399 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/go-github/github"
+	"github.com/mattermost/mattermost-server/mlog"
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/plugin"
+
+	"golang.org/x/oauth2"
+)
+
+const (
+	API_ERROR_ID_NOT_CONNECTED = "not_connected"
+	GITHUB_ICON_URL            = "https://assets-cdn.github.com/images/modules/logos_page/GitHub-Mark.png"
+	GITHUB_USERNAME            = "GitHub Plugin"
+)
+
+type APIErrorResponse struct {
+	ID         string `json:"id"`
+	Message    string `json:"message"`
+	StatusCode int    `json:"status_code"`
+}
+
+func writeAPIError(w http.ResponseWriter, err *APIErrorResponse) {
+	b, _ := json.Marshal(err)
+	w.WriteHeader(err.StatusCode)
+	w.Write(b)
+}
+
+func (p *Plugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Request) {
+	if err := p.IsValid(); err != nil {
+		http.Error(w, "This plugin is not configured.", http.StatusNotImplemented)
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+
+	switch path := r.URL.Path; path {
+	case "/webhook":
+		p.handleWebhook(w, r)
+	case "/oauth/connect":
+		p.connectUserToGitHub(w, r)
+	case "/oauth/complete":
+		p.completeConnectUserToGitHub(w, r)
+	case "/api/v1/connected":
+		p.getConnected(w, r)
+	case "/api/v1/todo":
+		p.postToDo(w, r)
+	case "/api/v1/reviews":
+		p.getReviews(w, r)
+	case "/api/v1/mentions":
+		p.getMentions(w, r)
+	case "/api/v1/unreads":
+		p.getUnreads(w, r)
+	case "/api/v1/settings":
+		p.updateSettings(w, r)
+	default:
+		http.NotFound(w, r)
+	}
+}
+
+func (p *Plugin) connectUserToGitHub(w http.ResponseWriter, r *http.Request) {
+	userID := r.Header.Get("Mattermost-User-ID")
+	if userID == "" {
+		http.Error(w, "Not authorized", http.StatusUnauthorized)
+		return
+	}
+
+	conf := p.getOAuthConfig()
+
+	state := fmt.Sprintf("%v_%v", model.NewId(), userID)
+
+	p.API.KVSet(state, []byte(state))
+
+	url := conf.AuthCodeURL(state, oauth2.AccessTypeOffline)
+
+	http.Redirect(w, r, url, http.StatusFound)
+}
+
+func (p *Plugin) completeConnectUserToGitHub(w http.ResponseWriter, r *http.Request) {
+	ctx := context.Background()
+	conf := p.getOAuthConfig()
+
+	code := r.URL.Query().Get("code")
+	if len(code) == 0 {
+		http.Error(w, "missing authorization code", http.StatusBadRequest)
+		return
+	}
+
+	state := r.URL.Query().Get("state")
+
+	if storedState, err := p.API.KVGet(state); err != nil {
+		http.Error(w, "missing stored state", http.StatusBadRequest)
+		return
+	} else if string(storedState) != state {
+		http.Error(w, "invalid state", http.StatusBadRequest)
+		return
+	}
+
+	userID := strings.Split(state, "_")[1]
+
+	p.API.KVDelete(state)
```

Comment:
I'ts a rare event, but it is possible for a key not to be deleted because the user agent never makes it here. Might be a good idea to allow some kind of mechanism for cleaning up old keys.

Reply:
Yeah I was thinking about that from the general plugin perspective, might be nice to add APIs such as `KVStoreWithExpiry()` so that plugins don't have to worry about it